Mathlib/Combinatorics/Optimization/ValuedCSP.lean

```diff
@@ -0,0 +1,216 @@
+  /-- Arity of the function -/
+  k : ℕ
+  /-- Which cost function is instantiated -/
+  f : (Fin k → Γ.D) → C
+  /-- The cost function comes from the template -/
+  inΓ : ⟨k, f⟩ ∈ Γ.F
+  /-- Which variables are plugged as arguments to the cost function -/
+  app : Fin k → ι
+
+def valuedCspTerm_of_unary {Γ : ValuedCspTemplate C} {ι : Type _} {f₁ : Γ.D → C}
+    (ok : ⟨1, n1ary_of_unary f₁⟩ ∈ Γ.F) (i : ι) : ValuedCspTerm Γ ι :=
+  ⟨1, n1ary_of_unary f₁, ok, ![i]⟩
+
+def valuedCspTerm_of_binary {Γ : ValuedCspTemplate C} {ι : Type _} {f₂ : Γ.D → Γ.D → C}
+    (ok : ⟨2, n2ary_of_binary f₂⟩ ∈ Γ.F) (i j : ι) : ValuedCspTerm Γ ι :=
+  ⟨2, n2ary_of_binary f₂, ok, ![i, j]⟩
+
+/-- Evaluation of a `Γ` term `t` for given solution `x`. -/
+def ValuedCspTerm.evalSolution {Γ : ValuedCspTemplate C} {ι : Type _}
+    (t : ValuedCspTerm Γ ι) (x : ι → Γ.D) : C :=
+  t.f (x ∘ t.app)
+
+/-- A valued CSP instance over the template `Γ` with variables indexed by `ι`.-/
+def ValuedCspInstance (Γ : ValuedCspTemplate C) (ι : Type _) : Type :=
+  List (ValuedCspTerm Γ ι)
+
+/-- Evaluation of a `Γ` instance `I` for given solution `x`. -/
+def ValuedCspInstance.evalSolution {Γ : ValuedCspTemplate C} {ι : Type _}
+    (I : ValuedCspInstance Γ ι) (x : ι → Γ.D) : C :=
+  (I.map (fun t => t.evalSolution x)).sum
+
+/-- Condition for `x` being an optimum solution (min) to given `Γ` instance `I`.-/
+def ValuedCspInstance.optimumSolution {Γ : ValuedCspTemplate C} {ι : Type _}
+    (I : ValuedCspInstance Γ ι) (x : ι → Γ.D) : Prop :=
+  ∀ y : ι → Γ.D, I.evalSolution x ≤ I.evalSolution y
+
+
+
+
+
+
+
+
+
+
+
+-- Example: minimize |x| + |y| where x and y are rational numbers
+
+private def absRat : (Fin 1 → ℚ) → ℚ := @n1ary_of_unary ℚ ℚ Abs.abs
+
+private def exampleAbs : Σ (k : ℕ), (Fin k → ℚ) → ℚ := ⟨1, absRat⟩
+
+private def exampleFiniteValuedCsp : ValuedCspTemplate ℚ :=
+  ValuedCspTemplate.mk ℚ {exampleAbs}
+
+private lemma abs_in : ⟨1, absRat⟩ ∈ exampleFiniteValuedCsp.F := rfl
+
+private def exampleFiniteValuedInstance : ValuedCspInstance exampleFiniteValuedCsp (Fin 2) :=
+  [valuedCspTerm_of_unary abs_in 0, valuedCspTerm_of_unary abs_in 1]
+
+#eval exampleFiniteValuedInstance.evalSolution ![(3 : ℚ), (-2 : ℚ)]
+
+example : exampleFiniteValuedInstance.optimumSolution ![(0 : ℚ), (0 : ℚ)] := by
+  unfold ValuedCspInstance.optimumSolution
+  unfold exampleFiniteValuedCsp
+  intro s
+  convert_to 0 ≤ ValuedCspInstance.evalSolution exampleFiniteValuedInstance s
+  rw [ValuedCspInstance.evalSolution, exampleFiniteValuedInstance,
+      List.map_cons, List.map_cons, List.map_nil, List.sum_cons, List.sum_cons, List.sum_nil,
+      add_zero]
+  show 0 ≤ |s 0| + |s 1|
+  have s0nn : 0 ≤ |s 0|
+  · exact abs_nonneg (s 0)
+  have s1nn : 0 ≤ |s 1|
+  · exact abs_nonneg (s 1)
+  linarith
+
+
+
+private def Bool_add_le_add_left (a b : Bool) :
+  (a = false ∨ b = true) → ∀ (c : Bool), (((c || a) = false) ∨ ((c || b) = true)) :=
+by simp
+
+-- Upside down !!
```

Comment:
What does this comment mean?

Reply:
It means that, in this example (which is here only temporarily and won't be added to mathlib), `True` and `False` are swapped out of convenience.

Reply:
I don't see the swap. Either way, this instance is bad because mathlib already defines `b1 + b2` as `b1 ^^^ b2`, but you define it with `||` instead.

( I realize it's just here for your example, but I think this makes it a bad example too)

Reply:
I should have defined an instance that makes `True ≤ False` and so on.

Reply:
If there is already an instance that treats plus as xor, it wouldn't be a linear order, right?
My code reuses `Bool.linearOrder` tho.